openwisp_controller/config/tasks.py

```diff
@@ -157,3 +158,46 @@ def invalidate_device_checksum_view_cache(organization_id):
         Device.objects.filter(organization_id=organization_id).only('id').iterator()
     ):
         DeviceChecksumView.invalidate_get_device_cache(device)
+
+
+@shared_task(soft_time_limit=7200)
+def fetch_whois_details(device_pk, ip):
+    """
+    Fetches the WHOIS details of the given IP address
+    and creates/updates the WHOIS record.
+    """
+    WHOISInfo = load_model('config', 'WHOISInfo')
+    try:
+        # 'geolite.info' host is used for GeoLite2
+        ip_client = geoip2.webservice.Client(
+            settings.GEOIP_ACCOUNT_ID, settings.GEOIP_LICENSE_KEY, 'geolite.info'
+        )
+
+        data = ip_client.city(ip)
+        # Format address using the data from the geoip2 response
+        address = ', '.join(
+            [
+                data.city.name,
+                data.country.name,
+                data.continent.name,
+                str(data.postal.code),
+            ]
+        )
```

Comment:
Should we concatenate all of this into a single field, or store each component in separate fields? If adding more fields isn’t ideal, we could consider using a jsonb field instead.

By the way, what would be the maximum length after concatenation?